src/main/java/com/cleanroommc/groovyscript/compat/mods/astralsorcery/Grindstone.java

```diff
@@ -0,0 +1,111 @@
+package com.cleanroommc.groovyscript.compat.mods.astralsorcery;
+
+import com.cleanroommc.groovyscript.api.GroovyBlacklist;
+import com.cleanroommc.groovyscript.api.GroovyLog;
+import com.cleanroommc.groovyscript.compat.mods.ModSupport;
+import com.cleanroommc.groovyscript.helper.ingredient.OreDictIngredient;
+import com.cleanroommc.groovyscript.helper.recipe.AbstractRecipeBuilder;
+import com.cleanroommc.groovyscript.registry.VirtualizedRegistry;
+import hellfirepvp.astralsorcery.common.crafting.ItemHandle;
+import hellfirepvp.astralsorcery.common.crafting.grindstone.GrindstoneRecipe;
+import hellfirepvp.astralsorcery.common.crafting.grindstone.GrindstoneRecipeRegistry;
+import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.ApiStatus;
+
+import java.util.ArrayList;
+
+public class Grindstone extends VirtualizedRegistry<GrindstoneRecipe> {
+
+    @Override
+    @GroovyBlacklist
+    @ApiStatus.Internal
+    public void onReload() {
+        removeScripted().forEach(this::remove);
+        restoreFromBackup().forEach(this::add);
+    }
+
+    private void add(GrindstoneRecipe recipe) {
+        GrindstoneRecipeRegistry.registerGrindstoneRecipe(recipe);
+    }
+
+    public GrindstoneRecipe add(ItemHandle input, ItemStack output, int chance, float doubleChance) {
+        GrindstoneRecipe recipe = new GrindstoneRecipe(input, output, chance, doubleChance);
+        addScripted(recipe);
+        return GrindstoneRecipeRegistry.registerGrindstoneRecipe(recipe);
+    }
+
+    private void remove(GrindstoneRecipe recipe) {
+        GrindstoneRecipeRegistry.recipes.remove(recipe);
+    }
+
+    public void removeByOutput(OreDictIngredient ore) {
+        for (ItemStack item : ore.getMatchingStacks())
+            this.removeByOutput(item);
+    }
+
+    public void removeByOutput(ItemStack item) {
+        ArrayList<GrindstoneRecipe> remove = new ArrayList<>();
+
+        for (GrindstoneRecipe recipe : GrindstoneRecipeRegistry.recipes) {
+            if (recipe.isValid() && recipe.getOutputForMatching().isItemEqual(item)) {
+                addBackup(recipe);
+                remove.add(recipe);
+            }
+        }
+
+        remove.forEach(recipe -> GrindstoneRecipeRegistry.recipes.remove(recipe));
+    }
+
+    public static RecipeBuilder recipeBuilder() {
+        return new RecipeBuilder();
+    }
+
+    public static class RecipeBuilder extends AbstractRecipeBuilder<GrindstoneRecipe> {
+
+        private ItemHandle input = null;
+        private int chance = 0;
+        private float doubleChance = 0.0F;
```

Comment:
Why are there two chances

Reply:
chance = chance that a single output is produced (sometimes you grind something and get nothing)
doubleChance = percentage chance of an additional output given a single output would be produced

Reply:
then why is it int and float? :concern:
and naming them primaryChance and secondaryChance makes more sense imo